Source/EbookShop.Services/UserService.cs

```diff
@@ -42,5 +43,94 @@ public async Task<DashboardDTO> IsValidUserHTTP()
+
+            Ebook[] temp = GetEbooksByTitle(context, text);
+            returnValue.AddRange(temp);
+
+            temp = context.Ebooks.Where(
+                (e) =>
+                    DoesEbookContainCategory(e, text)
+            ).ToArray();
+            returnValue.AddRange(temp);
+
+            temp = context.Ebooks.Where(
+                (e) =>
+                    DoesEbookContainAuthor(e, text)
+            ).ToArray();
+            returnValue.AddRange(temp);
+
+            return returnValue.ToArray();
+        }
+
+        //Zwraca tablicę z ebookami, w których tytule zawiera się podana fraza
+        public static Ebook[] GetEbooksByTitle(EbookShopContext context, string title)
+        {
+            Ebook[] returnValue = context.Ebooks.Where(e => e.Title.Contains(title)).ToArray();
+
+            return returnValue;
+        }
+
+        //Zwraca tablicę ebooków danej kategorii
+        public static Ebook[] GetEbooksByCategory(EbookShopContext context, Category category)
+        {
+            Ebook[] returnValue = context.Ebooks.Where(
+                (e) =>
+                    DoesEbookContainCategory(e, category)
+            ).ToArray();
+
+            return returnValue;
+        }
+
+        //Sprawdza czy ebook jest danej kategorii
+        private static bool DoesEbookContainCategory(Ebook e, Category category)
+        {
+            foreach (EbookCategories ebookCategories in e.EbookCategories)
+                return ebookCategories.Category == category;
+
+            return false;
+        }
+
+        //Sprawdza czy dana fraza zawiera się w nazwie kategorii ebooka
+        private static bool DoesEbookContainCategory(Ebook e, string category)
+        {
+            foreach (EbookCategories ebookCategories in e.EbookCategories)
+                return ebookCategories.Category.CategoryName.Contains(category);
+
+            return false;
+        }
+
+        //Zwraca tablicę ebooków z danym autorem
+        public static Ebook[] GetEbooksByAuthor(EbookShopContext context, Author author)
+        {
+            Ebook[] returnValue = context.Ebooks.Where(
+                (e) =>
+                    DoesEbookContainAuthor(e, author)
+            ).ToArray();
+
+            return returnValue;
+        }
+
+        //Sprawdza czy dany autor jest autorem danego ebooka
+        private static bool DoesEbookContainAuthor(Ebook e, Author author)
+        {
+            foreach (AuthorEbooks ebookCategories in e.AuthorEbooks)
+                return ebookCategories.Author == author;
+
+            return false;
+        }
+
+        //Sprawdza czy dana fraza zawiera się w imieniu bądź nazwisku autora
+        private static bool DoesEbookContainAuthor(Ebook e, string author)
+        {
+            foreach (AuthorEbooks ebookCategories in e.AuthorEbooks)
+                return ebookCategories.Author.FirstName.Contains(author) || ebookCategories.Author.LastName.Contains(author);
```

Comment:
To chyba trzeba dodatkowo obsłużyć bo zwróci wynik tylko pierwszej iteracji foreach-a?